assignment-client/src/audio/AudioMixerSlavePool.h

```diff
@@ -22,39 +22,75 @@
+// what information is needs to be thread-safe
+struct AudioMixerWorkerPoolData {
+    using Queue = tbb::concurrent_queue<SharedNodePointer>;
+    using Mutex = std::mutex;
+    using ConditionVariable = std::condition_variable;
+
+    // synchronization state
+    Mutex poolMutex;                   // only used for _poolCondition at the moment
+    ConditionVariable poolCondition;   // woken when work has been completed (_numStarted = _numFinished = _numThreads)
+    Mutex workerMutex;                 // only used for _workerCondition at the moment
+    ConditionVariable workerCondition; // woken when work needs to be done (_numStarted < _numThreads)
+
+    // The variables below this point are alternately owned by the pool or by the workers collectively.
+    // When idle they are owned by the pool.
+    // Moving ownership to the workers is done by setting _numStarted = _numFinished = 0 and waking _workerCondition
+    // Moving ownership to the pool is done when _numFinished == _numThreads and is done by waking _poolCondition
+
+    void (AudioMixerSlave::*function)(const SharedNodePointer& node);  // r/o when owned by workers, r/w when owned by pool
+    std::function<void(AudioMixerSlave&)> configure;  // r/o when owned by workers, r/w when owned by pool
+
+    // Number of currently-running worker threads
+    // r/o when owned by workersves, r/w when owned by pool
+    int numThreads{ 0 };
+
+    // Number of worker threads "awake" and processing the current request (0 <= _numStarted <= _numThreads)
+    // incremented when owned by workers, r/w when owned by pool
+    std::atomic<int> numStarted{ 0 };
+
+    // Number of worker threads finished with the current request (0 <= _numFinished <= _numStarted)
+    // incremented when owned by workers, r/w when owned by pool
+    std::atomic<int> numFinished{ 0 };
 
-class AudioMixerSlaveThread : public QThread, public AudioMixerSlave {
+    // Number of worker threads shutting down when asked to (0 <= _numStopped <= _numFinished)
+    // incremented when owned by workers, r/w when owned by pool
+    std::atomic<int> numStopped{ 0 };
+
+    // frame state
+    Queue queue;
+};
+
+class AudioMixerWorkerThread : public QThread, public AudioMixerSlave {
     Q_OBJECT
     using ConstIter = NodeList::const_iterator;
     using Mutex = std::mutex;
     using Lock = std::unique_lock<Mutex>;
 
 public:
-    AudioMixerSlaveThread(AudioMixerSlavePool& pool, AudioMixerSlave::SharedData& sharedData)
+    AudioMixerWorkerThread(AudioMixerWorkerPoolData& pool, AudioMixerSlave::SharedData& sharedData)
         : AudioMixerSlave(sharedData), _pool(pool) {}
 
     void run() override final;
+    inline void stop() { _stop = true; }
 
 private:
-    friend class AudioMixerSlavePool;
-
     void wait();
     void notify(bool stopping);
     bool try_pop(SharedNodePointer& node);
 
-    AudioMixerSlavePool& _pool;
+    AudioMixerWorkerPoolData& _pool;
```

Comment:
maybe _poolData?

Reply:
Whatever it is should be short, it's _data in the pool class and _pool here, and is a qualifier in front of all field accesses.

Reply:
I’m fine with it being _data in both places